yb-voyager/src/tgtdb/oracle.go

```diff
@@ -0,0 +1,478 @@
+	var outbuf bytes.Buffer
+	var errbuf bytes.Buffer
+	cmd.Stdout = &outbuf
+	cmd.Stderr = &errbuf
+	// fmt.Printf("cmd: %v\n", cmd)
+	err = cmd.Run()
+	log.Infof("sqlldr output: %s", outbuf.String())
+	log.Errorf("sqlldr error: %s", errbuf.String())
+	var err2 error
+	rowsAffected, err2 = getRowsAffected(outbuf.String())
+	if err2 != nil {
+		return 0, fmt.Errorf("get rows affected from sqlldr output: %w", err)
+	}
+
+	if err != nil {
+		// find ORA-00001: unique constraint * violated in log file
+		pattern := regexp.MustCompile(`ORA-00001: unique constraint \(.+?\) violated`)
+		scanner := bufio.NewScanner(sqlldrLogFile)
+		for scanner.Scan() {
+			line := scanner.Text()
+			if pattern.MatchString(line) {
+				err = db.recordEntryInDB(tx, batch, rowsAffected)
+				if err != nil {
+					err = fmt.Errorf("record entry in DB for batch %q: %w", batch.GetFilePath(), err)
+				}
+				return 0, err
+			}
+		}
+
+		return rowsAffected, fmt.Errorf("run sqlldr: %w", err)
+	}
+
+	err = db.recordEntryInDB(tx, batch, rowsAffected)
+	if err != nil {
+		err = fmt.Errorf("record entry in DB for batch %q: %w", batch.GetFilePath(), err)
+	}
+
+	return rowsAffected, err
+}
+
+func (db *TargetOracleDB) recordEntryInDB(tx *sql.Tx, batch Batch, rowsAffected int64) error {
+	cmd := batch.GetQueryToRecordEntryInDB(db.tconf.TargetDBType, rowsAffected)
+	_, err := tx.ExecContext(context.Background(), cmd)
+	if err != nil {
+		fmt.Printf("Error running query: %v\n", err)
+		return fmt.Errorf("insert into %s: %w", BATCH_METADATA_TABLE_NAME, err)
+	}
+	return nil
+}
+
+func getRowsAffected(outbuf string) (int64, error) {
+	regex := regexp.MustCompile(`Load completed - logical record count (\d+).`)
+	matches := regex.FindStringSubmatch(outbuf)
+	if len(matches) < 2 {
+		return 0, fmt.Errorf("no rows affected found in the sqlldr output")
+	}
+	return strconv.ParseInt(matches[1], 10, 64)
+}
+
+func getValue(connectionString, key string) string {
+	startIndex := strings.Index(connectionString, key)
+	if startIndex == -1 {
+		return ""
+	}
+
+	startIndex += len(key) + 2
+	endIndex := strings.Index(connectionString[startIndex:], "\"")
+
+	return connectionString[startIndex : startIndex+endIndex]
+}
+
+func (db *TargetOracleDB) isBatchAlreadyImported(tx *sql.Tx, batch Batch) (bool, int64, error) {
+	var rowsImported int64
+	query := batch.GetQueryIsBatchAlreadyImported(db.tconf.TargetDBType)
+	err := tx.QueryRowContext(context.Background(), query).Scan(&rowsImported)
+	if err == nil {
+		log.Infof("%v rows from %q are already imported", rowsImported, batch.GetFilePath())
+		return true, rowsImported, nil
+	}
+	if err == sql.ErrNoRows {
+		log.Infof("%q is not imported yet", batch.GetFilePath())
+		return false, 0, nil
+	}
+	return false, 0, fmt.Errorf("check if %s is already imported: %w", batch.GetFilePath(), err)
+}
+
+func (db *TargetOracleDB) setTargetSchema(conn *sql.Conn) {
+	checkSchemaExistsQuery := fmt.Sprintf(
+		"SELECT 1 FROM ALL_USERS WHERE USERNAME = '%s'",
```

Comment:
For each USER there is an identically named SCHEMA. But is the reverse true? Can there be a SCHEMA without any similarly named USER?